source/extensions/quic_listeners/quiche/platform/quic_text_utils_impl.h

```diff
@@ -0,0 +1,128 @@
+#pragma once
+
+#include "absl/strings/ascii.h"
+#include "absl/strings/escaping.h"
+#include "absl/strings/match.h"
+#include "absl/strings/numbers.h"
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_format.h"
+#include "absl/strings/str_split.h"
+#include "quiche/quic/platform/api/quic_string.h"
+#include "quiche/quic/platform/api/quic_string_piece.h"
+
+// NOLINT(namespace-envoy)
+
+// This file is part of the QUICHE platform implementation, and is not to be
+// consumed or referenced directly by other Envoy code. It serves purely as a
+// porting layer for QUICHE.
+
+namespace quic {
+
+class QuicTextUtilsImpl {
+public:
+  static bool StartsWith(QuicStringPiece data, QuicStringPiece prefix) {
+    return absl::StartsWith(data, prefix);
+  }
+
+  static bool EndsWithIgnoreCase(QuicStringPiece data, QuicStringPiece suffix) {
+    return absl::EndsWithIgnoreCase(data, suffix);
+  }
+
+  static QuicString ToLower(QuicStringPiece data) { return absl::AsciiStrToLower(data); }
+
+  static void RemoveLeadingAndTrailingWhitespace(QuicStringPiece* data) {
+    size_t count = 0;
+    const char* ptr = data->data();
+    while (count < data->size() && absl::ascii_isspace(*ptr)) {
+      ++count;
+      ++ptr;
+    }
+    data->remove_prefix(count);
+
+    count = 0;
+    ptr = data->data() + data->size() - 1;
+    while (count < data->size() && absl::ascii_isspace(*ptr)) {
+      ++count;
+      --ptr;
+    }
+    data->remove_suffix(count);
+  }
+
+  static bool StringToUint64(QuicStringPiece in, uint64_t* out) {
+    return absl::SimpleAtoi(in, out);
+  }
+
+  static bool StringToInt(QuicStringPiece in, int* out) { return absl::SimpleAtoi(in, out); }
+
+  static bool StringToUint32(QuicStringPiece in, uint32_t* out) {
+    return absl::SimpleAtoi(in, out);
+  }
+
+  static bool StringToSizeT(QuicStringPiece in, size_t* out) { return absl::SimpleAtoi(in, out); }
+
+  static QuicString Uint64ToString(uint64_t in) { return absl::StrCat(in); }
+
+  static QuicString HexEncode(QuicStringPiece data) { return absl::BytesToHexString(data); }
+
+  static QuicString Hex(uint32_t v) { return absl::StrCat(absl::Hex(v)); }
+
+  static QuicString HexDecode(QuicStringPiece data) { return absl::HexStringToBytes(data); }
+
+  static void Base64Encode(const uint8_t* data, size_t data_len, QuicString* output) {
+    absl::Base64Escape(QuicString(reinterpret_cast<const char*>(data), data_len), output);
```

Comment:
WebSafeBase64Escape seems to not do padding but does have slightly different rules.  Is it usable or no?

Reply:
I'm not sure it's safe to: QuicTextUtils::Base64Encode is used for generating client channel IDs, so the minor difference in characters used could actually have an effect. (That said, this is for channel ID, which last I heard is being (or maybe has been) deprecated.)

Ok to leave as-is for the time being, and then hopefully remove once channel ID code is removed from QUICHE core code?